hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlockBuilder.java

```diff
@@ -0,0 +1,108 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.io.hfile;
+
+import static javax.swing.Spring.UNSET;
+import static org.apache.hadoop.hbase.io.ByteBuffAllocator.HEAP;
+
+import org.apache.hadoop.hbase.io.ByteBuffAllocator;
+import org.apache.hadoop.hbase.nio.ByteBuff;
+import org.apache.yetus.audience.InterfaceAudience;
+
+@InterfaceAudience.Private
+public class HFileBlockBuilder {
+
+  private BlockType blockType;
+  private int onDiskSizeWithoutHeader;
+  private int onDiskDataSizeWithHeader;
+  private int uncompressedSizeWithoutHeader;
+  private long prevBlockOffset;
+  private ByteBuff buf;
+  private boolean fillHeader = false;
+  private long offset = UNSET;
+  private int nextBlockOnDiskSize = UNSET;
+  private HFileContext fileContext;
+  private ByteBuffAllocator allocator = HEAP;
+
+  public HFileBlockBuilder withBlockType(BlockType blockType) {
+    this.blockType = blockType;
+    return this;
+  }
+
+  public HFileBlockBuilder withOnDiskSizeWithoutHeader(int onDiskSizeWithoutHeader) {
+    this.onDiskSizeWithoutHeader = onDiskSizeWithoutHeader;
+    return this;
+  }
+
+  public HFileBlockBuilder withOnDiskDataSizeWithHeader(int onDiskDataSizeWithHeader) {
+    this.onDiskDataSizeWithHeader = onDiskDataSizeWithHeader;
+    return this;
+  }
+
+  public HFileBlockBuilder withUncompressedSizeWithoutHeader(int uncompressedSizeWithoutHeader) {
+    this.uncompressedSizeWithoutHeader = uncompressedSizeWithoutHeader;
+    return this;
+  }
+
+  public HFileBlockBuilder withPrevBlockOffset(long prevBlockOffset) {
+    this.prevBlockOffset = prevBlockOffset;
+    return this;
+  }
+
+  public HFileBlockBuilder withByteBuff(ByteBuff buf) {
+    this.buf = buf;
+    return this;
+  }
+
+  public HFileBlockBuilder withFillHeader(boolean fillHeader) {
+    this.fillHeader = fillHeader;
+    return this;
+  }
+
+  public HFileBlockBuilder withOffset(long offset) {
+    this.offset = offset;
+    return this;
+  }
+
+  public HFileBlockBuilder withNextBlockOnDiskSize(int nextBlockOnDiskSize) {
+    this.nextBlockOnDiskSize = nextBlockOnDiskSize;
+    return this;
+  }
+
+  public HFileBlockBuilder withHFileContext(HFileContext fileContext) {
+    this.fileContext = fileContext;
+    return this;
+  }
+
+  public HFileBlockBuilder withByteBuffAllocator(ByteBuffAllocator allocator) {
+    this.allocator = allocator;
+    return this;
+  }
+
+  public HFileBlock build() {
+    if (buf.hasArray()) {
```

Comment:
I still feel we should take any assumptions here.  Better keep the notion of whether the block is backed by an exclusive or shared memory area. Based on that create different type of object.  Here we always see whether on heap and if so exclusive memory. Why we have to drop that way?

Reply:
> Why we have to drop that way?

You mean MemoryType ?  It's a concept around all the IOEngine / BlockCache / ByteBuffAllocator /HFileBlock  etc... Actually,  as I commented above,  heap or offheap can decide the shared or non-shared.  Then all those MemoryType places can be uniformed in this build() method here,  I'd prefer this simple implementation.

Reply:
Dont mean to keep the memory Type. That is not needed in HFileBlock. Even in the Builder too..  A setter which takes boolean to say whether it is exclusive memory or not would have been enough. Any way this assumption is true as of now. So its ok a continue this way if u strongly feel so

Reply:
OK,  will do. Thanks.